service/history/tieredStorageQueueProcessor.go

```diff
@@ -0,0 +1,306 @@
+		return true, nil
+	}
+	maxReadAckLevel := func() int64 {
+		return shard.GetTransferMaxReadLevel()
+	}
+	updateTieredStorageAckLevel := func(ackLevel int64) error {
+		return shard.UpdateTieredStorageAckLevel(ackLevel)
+	}
+
+	tieredStorageQueueShutdown := func() error {
+		return nil
+	}
+
+	retProcessor := &tieredStorageQueueProcessorImpl{
+		shard:                       shard,
+		options:                     options,
+		maxReadAckLevel:             maxReadAckLevel,
+		updateTieredStorageAckLevel: updateTieredStorageAckLevel,
+		tieredStorageQueueShutdown:  tieredStorageQueueShutdown,
+		tieredStorageTaskFilter:     tieredStorageTaskFilter,
+		logger:                      logger,
+		metricsClient:               historyEngine.metricsClient,
+		taskExecutor: newTieredStorageQueueTaskExecutor(
+			shard,
+			historyEngine,
+			logger,
+			historyEngine.metricsClient,
+			config,
+		),
+
+		config:       config,
+		ackLevel:     shard.GetTieredStorageAckLevel(),
+		shutdownChan: make(chan struct{}),
+
+		queueAckMgr:        nil, // is set bellow
+		queueProcessorBase: nil, // is set bellow
+		executionManager:   shard.GetExecutionManager(),
+	}
+
+	queueAckMgr := newQueueAckMgr(
+		shard,
+		options,
+		retProcessor,
+		shard.GetTieredStorageAckLevel(),
+		logger,
+	)
+
+	queueProcessorBase := newQueueProcessorBase(
+		shard.GetService().GetClusterMetadata().GetCurrentClusterName(),
+		shard,
+		options,
+		retProcessor,
+		queueAckMgr,
+		historyEngine.historyCache,
+		logger,
+		shard.GetMetricsClient().Scope(metrics.TieredStorageQueueProcessorScope),
+	)
+	retProcessor.queueAckMgr = queueAckMgr
+	retProcessor.queueProcessorBase = queueProcessorBase
+
+	return retProcessor
+}
+
+func (t *tieredStorageQueueProcessorImpl) Start() {
+	if !atomic.CompareAndSwapInt32(&t.isStarted, 0, 1) {
+		return
+	}
+	t.queueProcessorBase.Start()
+	go t.completeTaskLoop()
+}
+
+func (t *tieredStorageQueueProcessorImpl) Stop() {
+	if !atomic.CompareAndSwapInt32(&t.isStopped, 0, 1) {
+		return
+	}
+	t.queueProcessorBase.Stop()
+	close(t.shutdownChan)
+}
+
+// NotifyNewTask - Notify the processor about the new tiered storage task arrival.
+// This should be called each time new tiered storage task arrives, otherwise tasks maybe delayed.
+func (t *tieredStorageQueueProcessorImpl) NotifyNewTask(
+	tieredStorageTasks []tasks.Task,
+) {
+	if len(tieredStorageTasks) != 0 {
+		t.notifyNewTask()
+	}
+}
+
+func (t *tieredStorageQueueProcessorImpl) completeTaskLoop() {
+	timer := time.NewTimer(t.config.TieredStorageProcessorCompleteTaskInterval())
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-t.shutdownChan:
+			// before shutdown, make sure the ack level is up to date
+			err := t.completeTask()
+			if err != nil {
+				t.logger.Error("Error complete tiered storage task", tag.Error(err))
+			}
+			return
+		case <-timer.C:
+			for attempt := 1; attempt <= t.config.TieredStorageProcessorCompleteTaskFailureRetryCount(); attempt++ {
```

Comment:
retrying on task range deletion may not be really necessary, but ok for now

Reply:
Going to keep it as is then.